File: src/main/java/com/smartwallet/watchlist/controller/WatchlistController.java
package com.smartwallet.watchlist.controller;

import com.smartwallet.dto.ApiResponse;
import com.smartwallet.entity.Asset;
import com.smartwallet.repository.AssetRepository;
import com.smartwallet.watchlist.entity.Watchlist;
import com.smartwallet.watchlist.entity.WatchlistItem;
import com.smartwallet.watchlist.repository.WatchlistItemRepository;
import com.smartwallet.watchlist.repository.WatchlistRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/v1/watchlist")
public class WatchlistController {

    private final WatchlistRepository watchlistRepository;
    private final WatchlistItemRepository itemRepository;
    private final AssetRepository assetRepository;

    public WatchlistController(WatchlistRepository watchlistRepository,
                          WatchlistItemRepository itemRepository,
                          AssetRepository assetRepository) {
        this.watchlistRepository = watchlistRepository;
        this.itemRepository = itemRepository;
        this.assetRepository = assetRepository;
    }

    private Long getCurrentUserId() {
        return 1L;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<Watchlist>>> getWatchlists() {
        Long userId = getCurrentUserId();
        List<Watchlist> watchlists = watchlistRepository.findByUserIdOrderByCreatedAtDesc(userId);
        return ResponseEntity.ok(ApiResponse.success(watchlists));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Watchlist>> createWatchlist(@RequestBody Map<String, String> body) {
        Long userId = getCurrentUserId();
        String name = body.get("name");
        
        Watchlist watchlist = new Watchlist();
        watchlist.setUserId(userId);
        watchlist.setName(name);
        
        watchlist = watchlistRepository.save(watchlist);
        return ResponseEntity.ok(ApiResponse.success(watchlist));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> deleteWatchlist(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        watchlistRepository.findByIdAndUserId(id, userId).ifPresent(w -> {
            itemRepository.deleteByWatchlistId(id);
            watchlistRepository.delete(w);
        });
        return ResponseEntity.ok(ApiResponse.success(null));
    }

    @GetMapping("/{id}/items")
    public ResponseEntity<ApiResponse<List<Asset>>> getWatchlistItems(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        List<WatchlistItem> items = itemRepository.findByWatchlistIdOrderByPositionAsc(id);
        
        List<Asset> assets = items.stream()
            .map(item -> assetRepository.findBySymbol(item.getAssetSymbol()).orElse(null))
            .filter(a -> a != null)
            .toList();
        
        return ResponseEntity.ok(ApiResponse.success(assets));
    }

    @PostMapping("/{id}/items")
    public ResponseEntity<ApiResponse<WatchlistItem>> addItem(@PathVariable Long id, @RequestBody Map<String, String> body) {
        String symbol = body.get("symbol").toUpperCase();
        
        if (itemRepository.existsByWatchlistIdAndAssetSymbol(id, symbol)) {
            return ResponseEntity.badRequest().build();
        }
        
        WatchlistItem item = new WatchlistItem();
        item.setWatchlistId(id);
        item.setAssetSymbol(symbol);
        
        long count = itemRepository.countByWatchlistId(id);
        item.setPosition((int) count + 1);
        
        item = itemRepository.save(item);
        return ResponseEntity.ok(ApiResponse.success(item));
    }

    @DeleteMapping("/{id}/items/{symbol}")
    public ResponseEntity<ApiResponse<Void>> removeItem(@PathVariable Long id, @PathVariable String symbol) {
        itemRepository.findByWatchlistIdAndAssetSymbol(id, symbol.toUpperCase())
            .ifPresent(itemRepository::delete);
        return ResponseEntity.ok(ApiResponse.success(null));
    }

    @PostMapping("/favorite/{symbol}")
    public ResponseEntity<ApiResponse<WatchlistItem>> addFavorite(@PathVariable String symbol) {
        Long userId = getCurrentUserId();
        Watchlist defaultList = watchlistRepository.findByUserIdAndIsDefaultTrue(userId)
            .orElseGet(() -> {
                Watchlist w = new Watchlist();
                w.setUserId(userId);
                w.setName("Favoritos");
                w.setDefault(true);
                return watchlistRepository.save(w);
            });
        
        WatchlistItem item = new WatchlistItem();
        item.setWatchlistId(defaultList.getId());
        item.setAssetSymbol(symbol.toUpperCase());
        item.setPosition(1);
        
        item = itemRepository.save(item);
        return ResponseEntity.ok(ApiResponse.success(item));
    }

    @DeleteMapping("/favorite/{symbol}")
    public ResponseEntity<ApiResponse<Void>> removeFavorite(@PathVariable String symbol) {
        Long userId = getCurrentUserId();
        watchlistRepository.findByUserIdAndIsDefaultTrue(userId).ifPresent(watchlist -> {
            itemRepository.findByWatchlistIdAndAssetSymbol(watchlist.getId(), symbol.toUpperCase())
                .ifPresent(itemRepository::delete);
        });
        return ResponseEntity.ok(ApiResponse.success(null));
    }

    @GetMapping("/favorites")
    public ResponseEntity<ApiResponse<List<Asset>>> getFavorites() {
        Long userId = getCurrentUserId();
        Optional<Watchlist> watchlistOpt = watchlistRepository.findByUserIdAndIsDefaultTrue(userId);
        
        if (watchlistOpt.isEmpty()) {
            return ResponseEntity.ok(ApiResponse.success(List.of()));
        }
        
        List<WatchlistItem> items = itemRepository.findByWatchlistIdOrderByPositionAsc(watchlistOpt.get().getId());
        List<Asset> assets = items.stream()
            .map(item -> assetRepository.findBySymbol(item.getAssetSymbol()).orElse(null))
            .filter(a -> a != null)
            .toList();
        
        return ResponseEntity.ok(ApiResponse.success(assets));
    }
}

File: src/main/kotlin/com/smartwallet/watchlist/web/WatchlistController.kt
package com.smartwallet.watchlist.web

import com.smartwallet.audit.service.AuditService
import com.smartwallet.common.ApiResponse
import com.smartwallet.watchlist.model.Alert
import com.smartwallet.watchlist.model.Watchlist
import com.smartwallet.watchlist.model.WatchlistItem
import com.smartwallet.watchlist.service.WatchlistService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

@RestController
@RequestMapping(value = ["/api", "/api/v1"])
class WatchlistController(
    private val watchlistService: WatchlistService,
    private val audit: AuditService
) {

    // Favorites
    @GetMapping("/watchlist/favorites")
    fun favorites(): ApiResponse<List<WatchlistItem>> =
        ApiResponse(data = watchlistService.favorites())

    @PostMapping("/watchlist/favorite/{symbol}")
    fun addFavorite(@PathVariable symbol: String): ApiResponse<Void> {
        watchlistService.addFavorite(symbol)
        audit.log(null, "ADD_FAVORITE", "Favorite", null)
        return ApiResponse(data = null)
    }

    @DeleteMapping("/watchlist/favorite/{symbol}")
    fun removeFavorite(@PathVariable symbol: String): ApiResponse<Void> {
        watchlistService.removeFavorite(symbol)
        audit.log(null, "REMOVE_FAVORITE", "Favorite", null)
        return ApiResponse(data = null)
    }

    // Watchlists
    @GetMapping("/watchlist")
    fun watchlists(): ApiResponse<List<Watchlist>> =
        ApiResponse(data = watchlistService.watchlists())

    @PostMapping("/watchlist")
    fun create(@RequestBody body: CreateWatchlistRequest): ApiResponse<Watchlist> =
        ApiResponse(data = watchlistService.createWatchlist(body.name)).also {
            audit.log(null, "CREATE_WATCHLIST", "Watchlist", it.data?.id)
        }

    @GetMapping("/watchlist/{id}/items")
    fun items(@PathVariable id: Long): ApiResponse<List<WatchlistItem>> =
        ApiResponse(data = watchlistService.items(id))

    @PostMapping("/watchlist/{id}/items")
    fun addItem(@PathVariable id: Long, @RequestBody body: AddWatchlistItemRequest): ApiResponse<WatchlistItem> =
        ApiResponse(data = watchlistService.addItem(id, body.symbol)).also {
            audit.log(null, "ADD_WATCHLIST_ITEM", "WatchlistItem", it.data?.id)
        }

    @DeleteMapping("/watchlist/{id}/items/{symbol}")
    fun removeItem(@PathVariable id: Long, @PathVariable symbol: String): ResponseEntity<Void> {
        watchlistService.removeItem(id, symbol)
        return ResponseEntity.noContent().build()
    }

    // Alerts
    @GetMapping("/alerts")
    fun alerts(): ApiResponse<List<Alert>> = ApiResponse(data = watchlistService.alerts())

    @PostMapping("/alerts")
    fun createAlert(@RequestBody body: CreateAlertRequest): ApiResponse<Alert> =
        ApiResponse(data = watchlistService.createAlert(body.symbol, body.alertType, body.targetValue))

    @DeleteMapping("/alerts/{id}")
    fun deleteAlert(@PathVariable id: Long): ResponseEntity<Void> {
        watchlistService.deleteAlert(id)
        audit.log(null, "DELETE_ALERT", "Alert", id)
        return ResponseEntity.noContent().build()
    }
}

data class CreateWatchlistRequest(val name: String)
data class AddWatchlistItemRequest(val symbol: String)
data class CreateAlertRequest(val symbol: String, val alertType: String, val targetValue: BigDecimal? = null)
